File: Runtime/Native/Android/Proxy/DeferredDeeplinkListenerProxy.cs
#if UNITY_ANDROID
using JetBrains.Annotations;
using UnityEngine;

namespace Io.AppMetrica.Native.Android.Proxy {
    internal class DeferredDeeplinkListenerProxy : AndroidJavaProxy {
        [NotNull]
        private readonly DeferredDeeplink.DeeplinkDelegate _onDeeplinkLoadedDelegate;
        [CanBeNull]
        private readonly DeferredDeeplink.ErrorDelegate _onErrorDelegate;

        public DeferredDeeplinkListenerProxy(
            [NotNull] DeferredDeeplink.DeeplinkDelegate onDeeplinkLoadedDelegate,
            [CanBeNull] DeferredDeeplink.ErrorDelegate onErrorDelegate
        ) : base("io.appmetrica.analytics.plugin.unity.DeferredDeeplinkListenerProxy") {
            _onDeeplinkLoadedDelegate = onDeeplinkLoadedDelegate;
            _onErrorDelegate = onErrorDelegate;
        }

        // ReSharper disable once InconsistentNaming
        public void onDeeplinkLoaded([NotNull] string deeplink) {
            _onDeeplinkLoadedDelegate(deeplink);
        }

        // ReSharper disable once InconsistentNaming
        public void onError([NotNull] string error, [NotNull] string referrer) {
            _onErrorDelegate?.Invoke(GetError(error), referrer == "<UNITY_NULL>" ? null : referrer);
        }

        private static DeferredDeeplink.Error GetError([NotNull] string str) {
            switch (str) {
                case "NoReferrer":
                    return DeferredDeeplink.Error.NoReferrer;
                case "NotAFirstLaunch":
                    return DeferredDeeplink.Error.NotAFirstLaunch;
                case "ParseError":
                    return DeferredDeeplink.Error.ParseError;
                default:
                    return DeferredDeeplink.Error.Unknown;
            }
        }
    }
}
#endif

File: Runtime/Plugins/Android/DeferredDeeplinkListenerProxy.java
package io.appmetrica.analytics.plugin.unity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;
import java.util.Map;
import io.appmetrica.analytics.DeferredDeeplinkListener;
import io.appmetrica.analytics.StartupParamsCallback;
import io.appmetrica.analytics.StartupParamsItem;
import io.appmetrica.analytics.StartupParamsItemStatus;

public interface DeferredDeeplinkListenerProxy {
    void onDeeplinkLoaded(@NonNull String deeplink);

    void onError(@NotNull String error, @NotNull String referrer);

    class Delegate implements DeferredDeeplinkListener {
        @NonNull
        private final DeferredDeeplinkListenerProxy proxy;

        public Delegate(@NonNull DeferredDeeplinkListenerProxy proxy) {
            this.proxy = proxy;
        }

        @Override
        public void onDeeplinkLoaded(@NonNull String deeplink) {
            proxy.onDeeplinkLoaded(deeplink);
        }

        @Override
        public void onError(@NonNull Error error, @Nullable String referrer) {
            proxy.onError(getErrorStr(error), referrer == null ? "" : referrer);
        }

        @NonNull
        private static String getErrorStr(@NonNull Error error) {
            switch (error) {
                case NO_REFERRER:
                    return "NoReferrer";
                case NOT_A_FIRST_LAUNCH:
                    return "NotAFirstLaunch";
                case PARSE_ERROR:
                    return "ParseError";
                default:
                    return "Unknown";
            }
        }
    }
}
